src/internal/parallel/each.go

```diff
@@ -0,0 +1,30 @@
+package parallel
+
+type Action func() error
+
+func ForEach(actions []Action, maxConcurrency int) []error {
```

Comment:
I propose replacing both this function and the Provider/ModuleList.Parallel function with the following
```
import (
	"golang.org/x/sync/errgroup"
)

func ParallelForEach[T any](elements []T, maxConcurrency int, f func(index int, element T) error) error {
	sem := make(chan struct{}, maxConcurrency)
	var g errgroup.Group
	for i, element := range elements {
		sem <- struct{}{}
		i, element := i, element
		g.Go(func() error {
			defer func() { <-sem }()
			return f(i, element)
		})
	}
	return g.Wait()
}
```

Reply:
I'd prefer to leave the Provider/ModuleList.Parallel functions as they are, since they also handle logging specific actions.  I'll take a look at refactoring those using your proposal.

Reply:
Sure

Reply:
It also looks like errgroup allows you to set a max concurrency internally, dramatically simplifying the code!

Reply:
Ah, TIL!

Reply:
Hmm, this also has the side effect of only returning the first error instead of a list of errors.  Is that preferred?

Reply:
I think that's fine, yeah.

Reply:
If that's the desired behavior, it becomes more complex as you would want to call WithContext(...) on the group with the command's context.

Reply:
Alright, I don't mind either way. Whatever is simpler.

Reply:
https://github.com/opentofu/registry-stable/issues/40